src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+                return ParseBalancingGrouping(
+                    openParenToken, questionToken,
+                    openToken, capture);
+            }
+
+            var closeToken = ParseCaptureGroupingCloseToken(ref openParenToken, openToken);
+
+            return new RegexCaptureGroupingNode(
+                openParenToken, questionToken,
+                openToken, capture, closeToken,
+                ParseGroupingEmbeddedExpression(_options), ParseGroupingCloseParen());
+        }
+
+        private RegexToken ParseCaptureGroupingCloseToken(ref RegexToken openParenToken, RegexToken openToken)
+        {
+            if ((openToken.Kind == RegexKind.LessThanToken && _currentToken.Kind == RegexKind.GreaterThanToken) ||
+                (openToken.Kind == RegexKind.SingleQuoteToken && _currentToken.Kind == RegexKind.SingleQuoteToken))
+            {
+                return _currentToken;
+            }
+
+            if (_currentToken.Kind == RegexKind.EndOfFile)
+            {
+                openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Unrecognized_grouping_construct,
+                    GetSpan(openParenToken, openToken)));
+            }
+            else
+            {
+                openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Invalid_group_name_Group_names_must_begin_with_a_word_character,
+                    _currentToken.GetSpan()));
+
+                // Rewind to where we were before seeing this bogus character.
+                _lexer.Position--;
+            }
+
+            return CreateMissingToken(
+                openToken.Kind == RegexKind.LessThanToken
+                    ? RegexKind.GreaterThanToken : RegexKind.SingleQuoteToken);
+        }
+
+        private RegexBalancingGroupingNode ParseBalancingGrouping(
+            RegexToken openParenToken, RegexToken questionToken,
+            RegexToken openToken, RegexToken firstCapture)
+        {
+            var minusToken = _currentToken;
+            var secondCapture = _lexer.TryScanNumberOrCaptureName();
+            if (secondCapture == null)
+            {
+                // Invalid group name: Group names must begin with a word character
+                ConsumeCurrentToken(allowTrivia: false);
+
+                openParenToken = openParenToken.AddDiagnosticIfNone(new EmbeddedDiagnostic(
+                    WorkspacesResources.Invalid_group_name_Group_names_must_begin_with_a_word_character,
+                    GetTokenSpanIncludingEOF(_currentToken)));
+
+                // If we weren't at the end of the text, go back to before whatever character
+                // we just consumed.
+                MoveBackBeforePreviousScan();
+                secondCapture = CreateMissingToken(RegexKind.CaptureNameToken);
+            }
+
+            var second = secondCapture.Value;
```

Comment:
Should we check that the first and second captures are the same kind? Or is it possible to have one be a number and the other be a name?

Reply:
Oooh.  Good catch!  I will have to check and see what happens in .net there.  and will ensure the same behavior in ours.

Reply:
What was the resolution?

---
In reply to: [178593785](https://github.com/dotnet/roslyn/pull/23984#discussion_r178593785) [](ancestors = 178593785)

Reply:
The resolutoin is that it is allowed.  There is a test for this here:

https://github.com/dotnet/roslyn/pull/25520/files#diff-c09d350e72f9ec3d676978c5c237e34bR7309

Reply:
Thanks

---
In reply to: [210053111](https://github.com/dotnet/roslyn/pull/23984#discussion_r210053111) [](ancestors = 210053111)